library/core/src/iter/adapters/array_chunks.rs

```diff
@@ -0,0 +1,146 @@
+use core::iter::FusedIterator;
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// An iterator that yields the elements of another iterator in
+/// chunks of size `N`.
+///
+/// This `struct` is created by the [`array_chunks`] method on [`Iterator`]. See
+/// its documentation for more.
+///
+/// [`array_chunks`]: Iterator::array_chunks
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+#[derive(Debug)]
+pub struct ArrayChunks<I: Iterator, const N: usize> {
+    iter: I,
+    buffer: [MaybeUninit<I::Item>; N],
```

Comment:
Maybe `Option<(Box<[MaybeUninit<I::Item>; N]>, usize)>` would make sense? Then nothing has to be allocated while the iterator is being passed around and re-wrapped in other iterators. This avoids bloat when N is large. Only when the end is reached and a size-mismatch occurs an allocation will be needed.

Or perhaps even a simple `Box<[T]>` will do in that case since it only has to be allocated once.

Reply:
This is in `core`, so allocation is not available. Maybe it could be moved to `alloc`, but I don't know if `Iterator` can be extended there. I could also make a struct that just holds the buffer and implement `IntoIterator` on a mutable reference to it. This would be similar to the `IntoChunks` adapter in `itertools`, but would be the first adapter in `core` to do this.